[WK2] constexpr-ify IPC encoding, decoding of std::tuple<> objects

zdobersek opened the pull request:
#### b807bfd9ce0a83f31884a9b093d37a127cca7d5a
<pre>
[WK2] constexpr-ify IPC encoding, decoding of std::tuple&lt;&gt; objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=246149">https://bugs.webkit.org/show_bug.cgi?id=246149</a>

Reviewed by Kimmo Kinnunen.

Simplify the ArgumentCoder&lt;std::tuple&lt;&gt;&gt; specialization, removing helper structs
and leaning into constexpr for a finer implementation.

The encoding remains the same, just the helper struct&apos;s methods inlined into the
ArgumentCoder specialization.

The decoding methods are reworked to use constexpr to make compile-time codepaths.
For each desired element, decoding is done into std::optional&lt;&gt; values that are
then passed through tuples of rvalue references into recursive calls, with the
index sequence adjusted each time to address the next element. Once all elements
are successfully decoded, the final tuple is constructed, with the decoded objects
all moved into that tuple. If decoding was unsuccessful at any point, a nullopt
value is returned immediately, as before.

This avoids concatenating tuples that was occurring for each decoded element, and
was ending up moving each contained element into the concatenation result for each
such concatenation.

* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::TupleEncoder::encode): Deleted.
(IPC::TupleDecoder&lt;&gt;::decode): Deleted.

Canonical link: <a href="https://commits.webkit.org/255496@main">https://commits.webkit.org/255496@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2ed65ac62afd15bdd6e323f422685af39f2d056

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92502 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1723 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23087 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102253 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1720 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30090 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84912 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98437 "Hash e2ed65ac for PR 5083 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98165 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1155 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79001 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28082 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83077 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82762 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71178 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36501 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16698 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34281 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17874 "Found 1 new test failure: compositing/backing/backing-store-attachment-animating-outside-viewport.html (failure)") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3808 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38150 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40488 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40059 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37028 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -519,48 +518,29 @@ struct TupleEncoder {
-    {
-        return TupleDecoder<Elements...>::decode(decoder);
+        if constexpr (Index < sizeof...(Elements)) {
+            std::optional<std::tuple_element_t<Index, std::tuple<Elements...>>> optional;
```
this could perhaps be `auto result = decoder.decode<Type>();`
IIRC this prevents the default or copy construction of Type. IIRC the case was that above would invoke the construction of the object inside optional with the actual object, when it did the spec-mandatory copy elision.
The details escape me, but contrasting:
`std::optional<Type> result = decoder.decode<Type>();` <-- needs less from Type
`std::optional<Type> result; result = decoder.decode<Type>();` <-- needs more from Type

So we would prefer needing less from Type.

You could name the variable as `result` or `value` instead of `optional`, as that would potentially clash with `std::optional` the type.

If you are interested and can compile for Mac, you can add a test case in TestIPC project, new a new  test file EncoderDecoderTest.cpp that invokes just the encoding/decoding. There you can create a test class that is not default constructible and ensure that the decoders / encoders keep working with such classes.

zdobersek replied:
I did try invoking `decoder.template decode<>()` but it got stuck on `Daemon::Decoder` using this ArgumentCoder specialization and not providing the method. But yes, it would be more efficient to use.

I'll land this and then enhance `Daemon::Decoder`.

I am building on Mac, I'll look into adding the unit tests too.

zdobersek committed:
[WK2] constexpr-ify IPC encoding, decoding of std::tuple<> objects
https://bugs.webkit.org/show_bug.cgi?id=246149

Reviewed by Kimmo Kinnunen.

Simplify the ArgumentCoder<std::tuple<>> specialization, removing helper structs
and leaning into constexpr for a finer implementation.

The encoding remains the same, just the helper struct's methods inlined into the
ArgumentCoder specialization.

The decoding methods are reworked to use constexpr to make compile-time codepaths.
For each desired element, decoding is done into std::optional<> values that are
then passed through tuples of rvalue references into recursive calls, with the
index sequence adjusted each time to address the next element. Once all elements
are successfully decoded, the final tuple is constructed, with the decoded objects
all moved into that tuple. If decoding was unsuccessful at any point, a nullopt
value is returned immediately, as before.

This avoids concatenating tuples that was occurring for each decoded element, and
was ending up moving each contained element into the concatenation result for each
such concatenation.

* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::TupleEncoder::encode): Deleted.
(IPC::TupleDecoder<>::decode): Deleted.

Canonical link: https://commits.webkit.org/255496@main
